daemon/api_systems.go

```diff
@@ -272,3 +328,218 @@ func postSystemActionInstall(c *Command, systemLabel string, req *systemActionRe
 		return BadRequest("unsupported install step %q", req.Step)
 	}
 }
+
+func parseValidationSets(validationSets []string) ([]*asserts.AtSequence, error) {
+	sets := make([]*asserts.AtSequence, 0, len(validationSets))
+	for _, vs := range validationSets {
+		account, name, seq, err := snapasserts.ParseValidationSet(vs)
+		if err != nil {
+			return nil, err
+		}
+
+		assertion := asserts.AtSequence{
+			Type:        asserts.ValidationSetType,
+			SequenceKey: []string{release.Series, account, name},
+			Pinned:      seq > 0,
+			Sequence:    seq,
+			Revision:    asserts.RevisionNotKnown,
+		}
+
+		sets = append(sets, &assertion)
+	}
+
+	return sets, nil
+}
+
+func readFormValue(form *Form, key string) (string, *apiError) {
+	values := form.Values[key]
+	if len(values) != 1 {
+		return "", BadRequest("expected exactly one %q value in form", key)
+	}
+	return values[0], nil
+}
+
+func readOptionalFormValue(form *Form, key string, defaultValue string) (string, *apiError) {
+	values := form.Values[key]
+	switch len(values) {
+	case 0:
+		return defaultValue, nil
+	case 1:
+		return values[0], nil
+	default:
+		return "", BadRequest("expected at most one %q value in form", key)
+	}
+}
+
+func readOptionalFormBoolean(form *Form, key string, defaultValue bool) (bool, *apiError) {
+	values := form.Values[key]
+	switch len(values) {
+	case 0:
+		return defaultValue, nil
+	case 1:
+		b, err := strconv.ParseBool(values[0])
+		if err != nil {
+			return false, BadRequest("cannot parse %q value as boolean: %s", key, values[0])
+		}
+		return b, nil
+	default:
+		return false, BadRequest("expected at most one %q value in form", key)
+	}
+}
+
+func postSystemActionCreateOffline(c *Command, form *Form) Response {
+	label, errRsp := readFormValue(form, "label")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	testSystem, errRsp := readOptionalFormBoolean(form, "test-system", false)
+	if errRsp != nil {
+		return errRsp
+	}
+
+	markDefault, errRsp := readOptionalFormBoolean(form, "mark-default", false)
+	if errRsp != nil {
+		return errRsp
+	}
+
+	vsetsList, errRsp := readOptionalFormValue(form, "validation-sets", "")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	var splitVSets []string
+	if vsetsList != "" {
+		splitVSets = strings.Split(vsetsList, ",")
+	}
+
+	// TODO: why is this a comma delimited list, rather than multiple values?
```

Comment:
I do not see any other instances, but I could have missed some. I just looked for us splitting on ",". I also don't have a strong preference, but I'll make it use multiple values in the form just so this isn't an odd feature of the API.

Reply:
After starting to do this, I bet the reason it was a comma-delimited list was so that the JSON and form APIs could use the same field names. If we allowed the form to take multiple "validation-set" fields, it might be confusing that the JSON version takes a (plural) "validation-sets" field.